Implement mate.academy courses parser with async optimization

vladyslav-tmf opened the pull request:
## Result:
![Screenshot_2](https://github.com/user-attachments/assets/1c85018c-6059-4f75-8519-d933e4ec429d)

vladyslav-tmf committed:
Implement course parser for mate.academy website

- Added requests and beautifulsoup4 to requirements.txt.
- Created get_page_content and parse_course functions to fetch and parse course data from mate.academy landing page.
- Updated .flake8 configuration to exclude all virtual environments.

vladyslav-tmf committed:
Implement optional task: parse modules and topics count

- Extended Course class with modules_count and topics_count fields.
- Added Selenium integration for dynamic content parsing.
- Implemented course details parsing with JS click handling.
- Added reliable topics count extraction from fact blocks.
- Updated main parsing function to include new course details.

vladyslav-tmf committed:
Add logging functionality to track parsing progress

- Added logging configuration with file and console output.
- Added progress tracking for main page parsing.
- Added detailed logging for course details extraction.
- Added informative messages about topics and modules count.
- Added summary logging for total courses processed.
- Updated .gitignore to exclude .log file.

vladyslav-tmf committed:
Optimize parsing performance with asyncio

- Converted synchronous code to asynchronous using asyncio and aiohttp.
- Implemented browser pool with semaphore for parallel processing.
- Added concurrent course details extraction.
- Updated test_parse.py to work with async testing.
- Added pytest-asyncio to requirements.txt.
- Improved execution time 5 times faster.

mateacademy-ai-mentor commented on app/parse.py:
```diff
@@ -1,12 +1,165 @@
+async def get_course_details(
+    course_url: str,
+    semaphore: asyncio.Semaphore,
+    driver_pool: list[webdriver.Chrome],
```
The type hint `list[webdriver.Chrome]` might cause issues in some Python versions. Consider using `List[webdriver.Chrome]` from the `typing` module for compatibility with older Python versions.

mateacademy-ai-mentor commented on app/parse.py:
```diff
@@ -1,12 +1,165 @@
+
+    async with semaphore:
+        # Get available driver from pool
+        driver = driver_pool.pop()
```
Popping an element from `driver_pool` without checking if the pool is empty might lead to an `IndexError`. Ensure that the pool is not empty before popping an element.

mateacademy-ai-mentor commented on app/parse.py:
```diff
@@ -1,12 +1,165 @@
+    )
+    logging.info(f"Found {len(course_cards)} courses to process")
+
+    driver_pool = [webdriver.Chrome() for _ in range(MAX_CONCURRENT_REQUESTS)]
```
Consider using a context manager or a try-except block to handle potential exceptions when initializing the `webdriver.Chrome` instances. This will ensure that resources are properly released in case of an error.

mateacademy-ai-mentor reviewed (approved):
Great job on implementing the task requirements and ensuring the core functionality works as expected! 🎉 While there are a few minor improvements suggested, such as handling exceptions with context managers or try-except blocks, checking the `driver_pool` before popping elements, and using compatible type hints, these do not impact the core functionality. Keep up the good work and consider these suggestions for future enhancements! 😊

✨ Please rate this review in the Platform's notification center. I will be grateful for your feedback! ✨

vladyslav-tmf committed:
Improve error handling and driver pool safety

- Added safe driver pool initialization with error handling.
- Added waiting mechanism for available drivers.
- Added specific Selenium exceptions handling.
- Added proper cleanup for Chrome drivers.
- Added detailed error logging.

vladyslav-tmf committed:
Fix formatting

vladyslav-tmf committed:
Add Chrome headless mode for CI environment

- Added Chrome options for headless mode.

vladyslav-tmf committed:
Simplify "Show more" button handling

- Simplified button selector by using tag name instead of class.
- Replaced JS click with direct Selenium click.

AnyoneClown reviewed (approved):
LGTM
